File: task1.py
def calculate_discount(cart_total, quantities):
    flat_10_discount = 10 if cart_total > 200 else 0
    bulk_5_discount = max([(quantity - 10) * 0.05 for quantity in quantities], default=0)
    bulk_10_discount = 10 if sum(quantities) > 20 else 0
    tiered_50_discount = 0

    if sum(quantities) > 30 and any(quantity > 15 for quantity in quantities):
        tiered_50_discount = sum([max(quantity - 15, 0) * 0.5 for quantity in quantities])

    discounts = {
        "flat_10_discount": flat_10_discount,
        "bulk_5_discount": bulk_5_discount,
        "bulk_10_discount": bulk_10_discount,
        "tiered_50_discount": tiered_50_discount
    }

    applied_discount = max(discounts, key=discounts.get)
    discount_amount = discounts[applied_discount]

    return applied_discount, discount_amount


def main():
    products = {"Product A": 20, "Product B": 40, "Product C": 50}
    quantities = []

    for product, price in products.items():
        quantity = int(input(f"Enter quantity for {product}: "))
        quantities.append(quantity)

    gift_wraps = [input(f"Is {product} wrapped as a gift? (yes/no): ").lower() == 'yes' for product in products]

    subtotal = sum([quantities[i] * products[list(products.keys())[i]] for i in range(len(products))])

    applied_discount, discount_amount = calculate_discount(subtotal, quantities)

    shipping_fee = (sum(quantities) // 10) * 5
    gift_wrap_fee = sum(gift_wraps)

    total = subtotal - discount_amount + shipping_fee + gift_wrap_fee

    for i in range(len(products)):
        print(f"{list(products.keys())[i]} - Quantity: {quantities[i]}, Total: {quantities[i] * products[list(products.keys())[i]]}")

    print(f"\nSubtotal: ${subtotal}")
    print(f"Discount Applied: {applied_discount}, Amount: ${discount_amount}")
    print(f"Shipping Fee: ${shipping_fee}")
    print(f"Gift Wrap Fee: ${gift_wrap_fee}")
    print(f"\nTotal: ${total}")


if _name_ == "_main_":
    main()

File: task1.js
function calculateDiscount(cartTotal, quantities) {
    const flat10Discount = cartTotal > 200 ? 10 : 0;
    const bulk5Discount = Math.max(...quantities.map(quantity => Math.max(quantity - 10, 0) * 0.05), 0);
    const bulk10Discount = sum(quantities) > 20 ? 10 : 0;
    let tiered50Discount = 0;

    if (sum(quantities) > 30 && quantities.some(quantity => quantity > 15)) {
        tiered50Discount = sum(quantities.map(quantity => Math.max(quantity - 15, 0) * 0.5));
    }

    const discounts = {
        flat_10_discount: flat10Discount,
        bulk_5_discount: bulk5Discount,
        bulk_10_discount: bulk10Discount,
        tiered_50_discount: tiered50Discount
    };

    const appliedDiscount = Object.keys(discounts).reduce((a, b) => discounts[a] > discounts[b] ? a : b);
    const discountAmount = discounts[appliedDiscount];

    return { appliedDiscount, discountAmount };
}

function sum(array) {
    return array.reduce((a, b) => a + b, 0);
}

function main() {
    const products = { "Product A": 20, "Product B": 40, "Product C": 50 };
    const quantities = [];

    for (const [product, price] of Object.entries(products)) {
        const quantity = parseInt(prompt(`Enter quantity for ${product}: `));
        quantities.push(quantity);
    }

    const giftWraps = Object.keys(products).map(product =>
        prompt(`Is ${product} wrapped as a gift? (yes/no): `).toLowerCase() === 'yes'
    );

    const subtotal = sum(quantities.map((quantity, i) => quantity * Object.values(products)[i]));

    const { appliedDiscount, discountAmount } = calculateDiscount(subtotal, quantities);

    const shippingFee = Math.floor(sum(quantities) / 10) * 5;
    const giftWrapFee = sum(giftWraps);

    const total = subtotal - discountAmount + shippingFee + giftWrapFee;

    for (let i = 0; i < Object.keys(products).length; i++) {
        const product = Object.keys(products)[i];
        console.log(`${product} - Quantity: ${quantities[i]}, Total: ${quantities[i] * Object.values(products)[i]}`);
    }

    console.log(`\nSubtotal: $${subtotal}`);
    console.log(`Discount Applied: ${appliedDiscount}, Amount: $${discountAmount}`);
    console.log(`Shipping Fee: $${shippingFee}`);
    console.log(`Gift Wrap Fee: $${giftWrapFee}\nTotal: $${total}`);
}

main();
